pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,309 @@
+	 *    Soft Temp Disable if CSE's Firmware SKU is Custom
+	 *
+	 */
+	meiDev, err := GetMeiPciDevice()
+	if err != nil {
+		return false, fmt.Errorf("failed to get PCI MEI device: %v", err)
+	}
+	log.Printf("MEI Device found: %s", meiDev)
+
+	// check that CSE's current working state is normal
+	cs, err := meiDev.ReadConfigRegister(0x40 /* PCI_ME_HFSTS1*/, 32)
+	if err != nil {
+		return false, fmt.Errorf("PCI config read failed: %v", err)
+	}
+	// check that the current working state is ME_HFS1_CWS_NORMAL (0x05) and
+	// current operation mode is ME_HFS1_COM_NORMAL (0x0).
+	// `working_state` is bits 1-4 and `operation_state` is bits 7-9.
+	if (cs&0xf) != MEHfs1CwsNormal || ((cs>>6)&0x7) != MEHfs1ComNormal {
+		return false, nil
+	}
+	// check that CSE's firmware SKU is not custom, and if it is, that the
+	// current operation mode is Soft Temp Disable.
+	cs, err = meiDev.ReadConfigRegister(0x60 /* PCI_ME_HFSTS3 */, 32)
+	if err != nil {
+		return false, fmt.Errorf("PCI config read failed: %v", err)
+	}
+	// fw_sku is in bits 5-7 . ME_HFS3_FW_SKU_CUSTOM is 0x5
+	if (cs>>4)&0x7 == MEHfs3FwSkuCustom {
+		// TODO implement the same as coreboot's cse_is_hfs1_com_soft_temp_disable()
+		return false, errors.New("IsHmrfpoEnableAllowed does not support checking for Soft Temp Disable yet")
+	}
+	return true, nil
+}
+
+type hmrfpoEnableMsg struct {
+	header MKHIHdr
+	nonce  uint32
+}
+
+func (hem hmrfpoEnableMsg) ToBytes() []byte {
+	var buf []byte
+	buf = append(buf, hem.header[:]...)
+	var nonce [4]byte
+	binary.LittleEndian.PutUint32(nonce[:], hem.nonce)
+	return append(buf, nonce[:]...)
+}
+
+type hmrfpoEnableResponse struct {
+	Header   MKHIHdr
+	FctBase  uint32
+	FctLimit uint32
+	Status   uint8
+	reserved [3]byte
+}
+
+func hmrfpoEnableResponseFromBytes(b []byte) (*hmrfpoEnableResponse, error) {
+	var resp hmrfpoEnableResponse
+	minlen := len(resp.Header)
+	maxlen := minlen +
+		4 /* FctBase */ +
+		4 /* FctLimit */ +
+		1 /* Status */ +
+		3 /* reserved bytes */
+	if len(b) != minlen && len(b) != maxlen {
+		return nil, fmt.Errorf("size mismatch, want %d/%d bytes, got %d", minlen, maxlen, len(b))
+	}
+	copy(resp.Header[:], b[:4])
+	if len(b) == minlen {
+		// don't parse the rest, we got a partial response
+		return &resp, nil
+	}
+	// TODO this could use u-root's pkg/uio
+	resp.FctBase = binary.LittleEndian.Uint32(b[4:8])
+	resp.FctLimit = binary.LittleEndian.Uint32(b[8:12])
+	resp.Status = b[12]
+	return &resp, nil
+}
+
+// HmrfpoEnable enables the HMRFPO (Host ME Region Flash Protection Override) via CSE,
+// see cse_hmrfpo_enable at
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L64
+func (m MKHIClient) HmrfpoEnable() error {
+
+	var hdr MKHIHdr
+	hdr.SetGroupID(MKHIGroupIDHmrfpo)
+	hdr.SetCommand(MKHIHmrfpoEnable)
+	canEnable, err := m.IsHmrfpoEnableAllowed()
+	if err != nil {
+		return fmt.Errorf("Enabling HMRFPO failed: %v", err)
```

Comment:
don't cap error strings: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
good catch, it was a `log.Printf` in a previous life